storage/shard.go

```diff
@@ -441,27 +473,70 @@ func (s *dbShard) Write(
 	annotation []byte,
 ) error {
 	// Prepare write
-	entry, err := s.writableSeries(id)
+	entry, opts, err := s.tryRetrieveWritableSeries(id)
 	if err != nil {
 		return err
 	}
 
-	// Perform write
-	err = entry.series.Write(ctx, timestamp, value, unit, annotation)
-	// Load series metadata before decrementing the writer count
-	// to ensure this metadata is snapshotted at a consistent state
-	seriesUniqueIndex := entry.index
-	seriesID := entry.series.ID()
-	entry.decrementWriterCount()
-	if err != nil {
-		return err
+	writable := entry != nil
+
+	// If no entry and we are not writing new series asynchronously
+	if !writable && !opts.writeNewSeriesAsync {
+		// Avoid double lookup by enqueueing insert immediately
+		result, err := s.enqueueInsertSeries(id, enqueueInsertOptions{
+			hasPendingWrite: false,
+		})
+		if err != nil {
+			return err
+		}
+		// Wait for the insert
+		result.wg.Wait()
+		// Retrieve the inserted entry
+		entry, err = s.writableSeries(id)
+		if err != nil {
+			return err
+		}
+		writable = true
+	}
+
+	var (
+		commitLogSeriesID          ts.ID
+		commitLogSeriesUniqueIndex uint64
+	)
+	if writable {
+		// Perform write
+		err = entry.series.Write(ctx, timestamp, value, unit, annotation)
+		// Load series metadata before decrementing the writer count
+		// to ensure this metadata is snapshotted at a consistent state
+		commitLogSeriesID = entry.series.ID()
```

Comment:
We explicitly don't pool IDs from series to protect against this very case, here is the comment from `series.go`:
```go
	// NB(r): We explicitly do not place this ID back into an
	// existing pool as high frequency users of series IDs such
	// as the commit log need to use the reference without the
	// overhead of ownership tracking.
	// Since series are purged so infrequently the overhead
	// of not releasing back an ID to a pool is amortized over
	// a long period of time.
```

I can add a similar comment here though, that's a good idea.